<launch>
    <master auto="start"/>
    <param name="/use_sim_time" value="true"/>

    <arg name="scenario_name" value="scen10"/>
    <arg name="scenario_type" value="low"/>
    <arg name="Nobst" value="10"/>
    
    <arg name="scenario_time" value="600"/>

    <arg name="seed" value="4"/>

    <!-- MAP -->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find mada)/world/$(arg scenario_name).yaml" respawn="false" />
    
    <!-- WORLD -->
    <node pkg="stage_ros" type="stageros" name="stageros" args="$(find mada)/world/$(arg scenario_name)_$(arg scenario_type).world" respawn="false">
        <param name="base_watchdog_timeout" value="0.2"/>
    </node>

    <!-- =================================================================================================================== -->
    <!--                                                    ROBOT                                                            -->
    <!-- =================================================================================================================== -->

    <!-- ********************************************************** -->
    <!-- ROBOT PARAMETERS -->
    <arg name="sX" value="0.3"/>
    <arg name="sY" value="0.3"/>
    <arg name="laser_range" value="7"/>
    <arg name="linearVel" value="0.55"/>
    <arg name="angularVel" value="1"/>
    <arg name="inflFact" value="1.0"/>

    <arg name="map_topic" value="map"/>

    <!-- 
    <arg name="robot_prefix" value=""/>
    -->
    <arg name="robot_prefix" value="robot_0/"/>

    <!-- <arg name="localization_topic" value="$(arg robot_prefix)base_pose_ground_truth"/> -->
    <arg name="laser_topic" value="$(arg robot_prefix)base_scan"/>
    <arg name="robot_frame" value="$(arg robot_prefix)base_link"/>
    <arg name="path_topic" value="$(arg robot_prefix)pathToFollow"/>
    <arg name="vel_topic" value="$(arg robot_prefix)cmd_vel"/>

    <arg name="occup_threshold" value="0.1"/>
    <arg name="forget_obstacles" value="2"/>
    <arg name="max_mission_time" value="$(arg scenario_time)"/>

    <arg name="monitoring_type" value="greedy"/>
    <arg name="navigation_type" value="pf"/>

    <arg name="mission_type" value="0"  doc="mission type [0: monitoring, 1: coverage]" />

    <arg name="result_filename" value="res.txt"/>

    <!-- ********************************************************** -->

    <!-- LOCALIZATION -->

    <!--
    
    AMCL
    <arg name="localization_topic" value="$(arg robot_prefix)amcl_pose"/>
    <include file="$(find mada)/launch/loc_amcl.xml">
        <arg name="scan_topic"      value="scan"/> 
        <arg name="initial_pose_x"  value="37.3"/>
        <arg name="initial_pose_y"  value="10.2"/>
        <arg name="initial_pose_a"  value="-1.57"/>
        <arg name="odom_frame_id"   value="$(arg robot_prefix)odom"/>
        <arg name="base_frame_id"   value="$(arg robot_prefix)base_footprint"/>
        <arg name="laser_topic" value="$(arg robot_prefix)base_scan"/>
        <arg name="localization_topic" value="$(arg localization_topic)"/>
    </include>

    FAKE LOCALIZATION
    <arg name="localization_topic" value="$(arg robot_prefix)amcl_pose"/>
    <node name="fake_localization" pkg="fake_localization" type="fake_localization" respawn="false">
        <remap from="amcl_pose" to="$(arg localization_topic)" />
        <remap from="scan_topic" to="$(arg laser_topic)" />
        <remap from="base_pose_ground_truth" to="$(arg robot_prefix)base_pose_ground_truth" />
        <param name="odom_frame_id"   value="$(arg robot_prefix)odom"/>
        <param name="base_frame_id"   value="$(arg robot_prefix)base_link"/>
        <param name="global_frame_id"   value="$(arg map_topic)"/>
    </node>

    ODOMETRY
    <arg name="localization_topic" value="$(arg robot_prefix)odom"/>
    <include file="$(find mada)/launch/loc_odom.xml">
        <arg name="map_topic"  value="$(arg map_topic)"/>
        <arg name="odom_frame_id"  value="$(arg robot_prefix)odom"/>
    </include>

    -->

    <arg name="localization_topic" value="$(arg robot_prefix)amcl_pose"/>
    <node name="fake_localization" pkg="fake_localization" type="fake_localization" respawn="false">
        <remap from="amcl_pose" to="$(arg localization_topic)" />
        <remap from="scan_topic" to="$(arg laser_topic)" />
        <remap from="base_pose_ground_truth" to="$(arg robot_prefix)base_pose_ground_truth" />
        <param name="odom_frame_id"   value="$(arg robot_prefix)odom"/>
        <param name="base_frame_id"   value="$(arg robot_prefix)base_link"/>
        <param name="global_frame_id"   value="$(arg map_topic)"/>
    </node>

    <!-- MADA NODE -->
    <node pkg="mada" type="$(arg monitoring_type)_$(arg navigation_type)" name="$(arg monitoring_type)_$(arg navigation_type)" output="screen" args="$(arg robot_prefix) $(arg map_topic) $(arg localization_topic) $(arg robot_frame) $(arg laser_topic) $(arg vel_topic) $(arg sX) $(arg sY) $(arg laser_range) $(arg linearVel) $(arg angularVel) $(arg occup_threshold) $(arg forget_obstacles) $(arg max_mission_time) $(arg inflFact) $(arg mission_type) $(arg result_filename)" />

    <!-- =================================================================================================================== -->
    <!--                                                    OBSTACLES                                                       -->
    <!-- =================================================================================================================== -->
    <!-- ********************************************************** -->
    <arg name="obstFootprint_topic" value="obstacle"/>
    <arg name="osX" value="0.25"/>
    <arg name="osY" value="0.45"/>
    <arg name="obstLinearVel" value="0.1833"/>
    <arg name="obstAngularVel" value="0.5"/>
    <arg name="obstLinearAcc" value="0.05"/>
    <arg name="obstAngularAcc" value="0.01"/>
    <arg name="navAreaType" value="0"/> <!-- 0: free movement, 1: navigation within partitions, 2: within circular areas -->
    <arg name="areas" value="[ 16.7 13 4 ; 9.5 6.5 3.2]"/>
    <!-- ********************************************************** -->

    <!-- args=[mapTopic footPrintTopic Nobst sX sY maxLinearVel maxAngularVel linearAcc angularAcc] -->
    <node pkg="mada" type="mada_obstacles" name="mada_obstacles" output="screen" args="$(arg map_topic) $(arg obstFootprint_topic) $(arg Nobst) $(arg osX) $(arg osY) $(arg obstLinearVel) $(arg obstAngularVel) $(arg obstLinearAcc) $(arg obstAngularAcc) $(arg seed) $(arg navAreaType) $(arg areas)"/>

    <!-- =================================================================================================================== -->

    <!-- RViZ -->
    <node name="rviz1" pkg="rviz" type="rviz" args="-d $(find mada)/rviz/$(arg scenario_name)/low_level_planning.rviz" />
    <node name="rviz2" pkg="rviz" type="rviz" args="-d $(find mada)/rviz/$(arg scenario_name)/high_level_planning.rviz" />

    <!-- SAVE DATA
    <node pkg="rosbag" type="record" name="record_all" output="screen" args="-O /home/yarik/rosbag/$(arg scenario_name)_test.bag $(arg vel_topic) $(arg localization_topic) /tf $(arg map_topic) /gradient /robot_ftprint /explorationMap /visibleDynObst /pathToGoal /totalPath /pathFollowerPlan /obstFootprints /robot_ftprint /segFront /targetPos"/> -->

</launch>



 
